### ISSUE TITLE
keyword-spacing does not work for function keyword

### ISSUE THREAD

**Issue by friederbluemle** [eslint/eslint#12523]
**Tell us about your environment**

Node version: v12.12.0
npm version: v6.11.3
Local ESLint version: v6.6.0 (Currently used)
Global ESLint version: Not found

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  "env": {
    "browser": true,
    "es6": true,
    "node": true
  },
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module"
  },
  "rules": {
    "keyword-spacing": "error",
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function test2(x) {
  return x;
}

const test1 = function(x) {
  if(true) {
    return 'test1' + x;
  } else {
    return 'test2' + x;
  }
};
```

```sh
eslint .
```

**What did you expect to happen?**

I expected **two errors** here

```
  5:15  error  Expected space(s) after "function"  keyword-spacing
  6:3   error  Expected space(s) after "if"  keyword-spacing
```

**What actually happened? Please include the actual, raw output from ESLint.**

Only **one** error is reported, and the missing space after the `function` keyword remains undetected:

```
$ eslint .

test.js
  6:3  error  Expected space(s) after "if"  keyword-spacing

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

Since the error is not detected, of course I'm also not able to automatically fix it (`--fix`).

**Are you willing to submit a pull request to fix this bug?**

I would, but not sure where to start looking...

**Comment by g-plane** [comment]
Hi, please use [space-before-function-paren](https://eslint.org/docs/rules/space-before-function-paren) instead.

**Comment by friederbluemle** [comment]
@g-plane - Thanks a lot for your comment and suggestion! I was able to get the desired behavior using `space-before-function-paren`.

However, that still leaves three open questions:

- Why does the `keyword-spacing` rule not detect the space?
- Does that mean there is some kind of internal "priority" (e.g. `keyword-spacing` has lower priority than `space-before-function-paren`)? If yes, where can I see this priority?
- The documentation lists `always` as the *default* for `space-before-function-paren`, though in order to make it do what I want, I had to **manually enable** and set it to:

```
'space-before-function-paren': [2, {"anonymous": "always", "named": "never", "asyncArrow": "always"}],
```

**Comment by g-plane** [comment]
- `space-before-function-paren` does it, so `keyword-spacing` doesn't need to detect again.
- No.
- What's your problem?

**Comment by platinumazure** [comment]
Hi @friederbluemle. I apologize on behalf of the team for letting this slip through the cracks. I will attempt to provide a more complete answer to your questions.

> - Why does the `keyword-spacing` rule not detect the space?

The short version is, because it intentionally is not designed to do so.

We had to make some decisions on how to make sure rules didn't conflict with each other. If `keyword-spacing` and `space-before-function-paren` both watched the same location, it would be possible for one rule to be configured to require a space, and the other rule to be configured to forbid a space. This would be a horrible user experience. So, in some cases, rules have specific cases they might not enforce, and instead would leave those cases to other rules.

In this case, `keyword-spacing` chooses to leave the space between `function` and parentheses up to `space-before-function-paren`.

I would expect/hope that the documentation for each of those rules to reference the other rule, under related rules at least. I would also hope the rule documentation would point out cases where other rules are responsible. If the documentation can be improved, please feel free to submit a pull request.

> - Does that mean there is some kind of internal "priority" (e.g. `keyword-spacing` has lower priority than `space-before-function-paren`)? If yes, where can I see this priority?

As I noted above, it's not a "priority" system and more that rules can collide in certain circumstances and we design the rules with some exceptions to avoid conflicts. So, `keyword-spacing` simply doesn't enforce that particular space.

- The documentation lists `always` as the *default* for `space-before-function-paren`, though in order to make it do what I want, I had to **manually enable** and set it to:

This is actually a different issue.

Rules are disabled by default. If you don't configure a rule (or extend a configuration which configures a rule), then that rule is completely disabled and will not run or report anything.

When a rule is enabled with no options (using syntax like `"space-before-function-paren": "error"` or `"space-before-function-paren": ["error"]`), then a rule will operate in a specific way by default. This is what the `space-before-function-paren` documentation was referring to when it said it uses the `"always"` option by default. `"space-before-function-paren": "error"` is the same as `"space-before-function-paren": ["error", "always"]`. However, you still need to explicitly configure the rule in your configuration in order to turn it on.

------

Also, I would like to apologize about the last response from a team member. It came off as rude, mostly due to language/cultural differences. I will work with that team member to explain that "What's your problem?" with no other context is a statement that can indicate strong objection, which was uncalled for here and which I don't believe was intended. He should have said something like "I don't understand the question. Can you please explain the problem in more detail?" Additionally, your second question deserved a more complete answer than "No", with more context in order to try to help you understand the underlying concepts.

I hope my answer has helped you and maybe partially made up for your earlier experience.

Thanks again for the issue. I apologize once more for both the unintentional rudeness and the delay in resolving this.

**Comment by friederbluemle** [comment]
Hi @platinumazure - Thank you very much for this very detailed response and explanation. The reasons given definitely make sense and I'm sure a lot of thought went into designing the system the way it is.

While I wouldn't say it's very intuitive for someone not knowing these details, I understand that's just how it works as of today, and as mentioned above, I was able to make it work for my use case. Maybe there's room for improvement on the documentation side at least.

Just as an idea (not sure if possible, since I don't have detailed knowledge of the system): If I understand correctly, `keyword-spacing` "knows" it may conflict with `space-before-function-paren` and therefore special-cases the `function` keyword. Would it be possible for `keyword-spacing` to _detect_ that `space-before-function-paren` is not even enabled/active in the current configuration, and in this case just treat all keywords equally?

And as a separate note (may require a separate issue/discussion): I don't think a simple on/off switch for `space-before-function-paren` is very useful (which is why I had to use the longish configuration line above). The reason is, almost all code formatter/IDEs follow these standard conventions (in their default configuration):

- Spaces around keywords
- **No space** between the **name** of a function and the opening parenthesis

Having a space between a named function and the opening paren is extremely uncommon.

Regarding spaces around keywords, the **one major outlier** here is `prettier`, which has a bug in the current version that _removes_ the space after the function keyword. This could be one of the reasons these ESLint rules are designed the way they are, and why we're seeing inconsistencies in a lot of repositories that use Prettier. I think Prettier will be fixed in the next major version, so this should improve over time.

**Comment by friederbluemle** [comment]
Oh, and regarding the perceived rudeness in the previous response - no worries at all. I could tell it was unintentional. Thanks again!

**Comment by kaicataldo** [comment]
> Would it be possible for keyword-spacing to detect that space-before-function-paren is not even enabled/active in the current configuration, and in this case just treat all keywords equally?

Rules don't know about each other by design, so this is unfortunately not currently possible.

> Regarding spaces around keywords, the one major outlier here is prettier, which has a bug in the current version that removes the space after the function keyword. This could be one of the reasons these ESLint rules are designed the way they are, and why we're seeing inconsistencies in a lot of repositories that use Prettier. I think Prettier will be fixed in the next major version, so this should improve over time.

I might be misunderstanding something, but are you using both ESLint and Prettier for code formatting? If so, I think it's best to turn off all style/formatting rules in ESLint and just let Prettier do its thing. There are a few ways to do this, as outlined in the [Prettier docs](https://prettier.io/docs/en/integrating-with-linters.html#eslint).
